### PR TITLE
Always return valid values from MemoryPoolIterator.PeekLong()

### PR DESCRIPTION
**Pull request by halter73**
- Fixed edge case where the iterator is at the end of a block
- Fixed edge case where one bits where improperly filled in on a right shift
- Make return value of PeekLong Nullable<ulong>

This fixes GetKnownMethod and GetKnownVersion

#1183

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs [hunk 1]
```diff
@@ -180,38 +180,46 @@ public int Peek()
...
             }
             else if (wasLastBlock)
             {
-                return -1;
+                return null;
+            }
+            else if (_block.End - _index == 0)
+            {
+                // This case can not fall through to the else block since it would cause a 64-bit right shift
+                // on blockLong which is equivalent to no shift at all instead of shifting in all zeros.
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs]: nit: add link to docs here

File: test/Microsoft.AspNetCore.Server.KestrelTests/MemoryPoolIteratorTests.cs [hunk 2]
```diff
@@ -491,7 +584,36 @@ public void GetsKnownVersion(string input, char endChar, bool expectedResult, st
             Assert.Equal(expectedResult, result);
             Assert.Equal(expectedKnownString, knownString);
 
+            // Test at boundary
+            var maxSplit = Math.Min(input.Length, 9);
+            var nextBlock = _pool.Lease();
+
+            for (var split = 0; split <= maxSplit; split++)
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/MemoryPoolIteratorTests.cs]: Can we have this as a param to the test? Loops in tests are not ideal.

**Comment by cesarbs** [comment]
Have you checked perf?

File: test/Microsoft.AspNetCore.Server.KestrelTests/MemoryPoolIteratorTests.cs (hunk 2 above)
Comment (halter73) [inline test/Microsoft.AspNetCore.Server.KestrelTests/MemoryPoolIteratorTests.cs]: I considered making an AtBlockBoundary() variation of this test but then there would be 8 times as many InlineData attributes as this test. I decided it wasn't worth it.

**Comment by halter73** [comment]
blockBytes = 4

    // ***** BenchmarkRunner: Finish  *****

    // * Export *
      BenchmarkDotNet.Artifacts\results\Program-report.csv
      BenchmarkDotNet.Artifacts\results\Program-report-github.md
      BenchmarkDotNet.Artifacts\results\Program-report.html

    // * Detailed results *
    Program_TestPeekLong
    Mean = 10.9378 ns, StdError = 0.1208 ns (1.1%); N = 94, StdDev = 1.1712 ns
    Min = 9.7801 ns, Q1 = 9.9900 ns, Median = 10.2942 ns, Q3 = 12.2092 ns, Max = 14.3264 ns
    IQR = 2.2192 ns, LowerFence = 6.6613 ns, UpperFence = 15.5379 ns
    ConfidenceInterval = [10.7010 ns; 11.1746 ns] (CI 95%)
    Skewness = 0.7567471859832, Kurtosis = 2.17264689458998


    Program_TestOldPeekLong
    Mean = 7.6406 ns, StdError = 0.0722 ns (0.94%); N = 152, StdDev = 0.8899 ns
    Min = 7.0658 ns, Q1 = 7.1958 ns, Median = 7.3198 ns, Q3 = 7.6299 ns, Max = 11.9282 ns
    IQR = 0.4341 ns, LowerFence = 6.5447 ns, UpperFence = 8.2810 ns
    ConfidenceInterval = [7.4991 ns; 7.7821 ns] (CI 95%)
    Skewness = 2.90174038803229, Kurtosis = 11.4208574606696

    Total time: 00:02:57 (177.76 sec)

    // * Summary *

    Host Process Environment Information:
    BenchmarkDotNet.Core=v0.9.9.0
    OS=Windows
    Processor=?, ProcessorCount=4
    Frequency=2435885 ticks, Resolution=410.5284 ns, Timer=TSC
    CLR=CORE, Arch=64-bit ? [RyuJIT]
    GC=Concurrent Workstation
    dotnet cli version: 1.0.0-preview3-003206

    Type=Program  Mode=Throughput

              Method |     Median |    StdDev |
    ---------------- |----------- |---------- |
        TestPeekLong | 10.2942 ns | 1.1712 ns |
     TestOldPeekLong |  7.3198 ns | 0.8899 ns |

    // ***** BenchmarkRunner: End *****
    BenchmarkDotNet.Reports.Summary

    ===============================================================================

    blockBytes = 8

    // ***** BenchmarkRunner: Finish  *****

    // * Export *
      BenchmarkDotNet.Artifacts\results\Program-report.csv
      BenchmarkDotNet.Artifacts\results\Program-report-github.md
      BenchmarkDotNet.Artifacts\results\Program-report.html

    // * Detailed results *
    Program_TestPeekLong
    Mean = 6.7445 ns, StdError = 0.0164 ns (0.24%); N = 120, StdDev = 0.1801 ns
    Min = 6.5244 ns, Q1 = 6.6375 ns, Median = 6.6939 ns, Q3 = 6.7817 ns, Max = 7.6369 ns
    IQR = 0.1442 ns, LowerFence = 6.4212 ns, UpperFence = 6.9979 ns
    ConfidenceInterval = [6.7123 ns; 6.7767 ns] (CI 95%)
    Skewness = 2.12975105177026, Kurtosis = 8.81764830860382


    Program_TestOldPeekLong
    Mean = 4.4123 ns, StdError = 0.0134 ns (0.3%); N = 150, StdDev = 0.1639 ns
    Min = 4.2170 ns, Q1 = 4.3362 ns, Median = 4.3703 ns, Q3 = 4.4244 ns, Max = 5.7321 ns
    IQR = 0.0883 ns, LowerFence = 4.2038 ns, UpperFence = 4.5568 ns
    ConfidenceInterval = [4.3860 ns; 4.4385 ns] (CI 95%)
    Skewness = 4.49365233725576, Kurtosis = 31.4237283182822


    Total time: 00:03:34 (214.88 sec)

    // * Summary *

    Host Process Environment Information:
    BenchmarkDotNet.Core=v0.9.9.0
    OS=Windows
    Processor=?, ProcessorCount=4
    Frequency=2435885 ticks, Resolution=410.5284 ns, Timer=TSC
    CLR=CORE, Arch=64-bit ? [RyuJIT]
    GC=Concurrent Workstation
    dotnet cli version: 1.0.0-preview3-003206

    Type=Program  Mode=Throughput

              Method |    Median |    StdDev |
    ---------------- |---------- |---------- |
        TestPeekLong | 6.6939 ns | 0.1801 ns |
     TestOldPeekLong | 4.3703 ns | 0.1639 ns |

    // ***** BenchmarkRunner: End *****
    BenchmarkDotNet.Reports.Summary

    ===============================================================================

    blockBytes = 8

    // ***** BenchmarkRunner: Finish  *****

    // * Export *
      BenchmarkDotNet.Artifacts\results\Program-report.csv
      BenchmarkDotNet.Artifacts\results\Program-report-github.md
      BenchmarkDotNet.Artifacts\results\Program-report.html

    // * Detailed results *
    Program_TestPeekLong
    Mean = 8.0800 ns, StdError = 0.0151 ns (0.19%); N = 110, StdDev = 0.1583 ns
    Min = 7.8461 ns, Q1 = 7.9674 ns, Median = 8.0208 ns, Q3 = 8.1384 ns, Max = 8.5859 ns
    IQR = 0.1710 ns, LowerFence = 7.7110 ns, UpperFence = 8.3949 ns
    ConfidenceInterval = [8.0504 ns; 8.1095 ns] (CI 95%)
    Skewness = 1.31255424107211, Kurtosis = 4.16615935723663


    Program_TestOldPeekLong
    Mean = 7.4274 ns, StdError = 0.0447 ns (0.6%); N = 143, StdDev = 0.5349 ns
    Min = 6.8275 ns, Q1 = 7.1446 ns, Median = 7.2402 ns, Q3 = 7.4688 ns, Max = 9.7341 ns
    IQR = 0.3241 ns, LowerFence = 6.6585 ns, UpperFence = 7.9549 ns
    ConfidenceInterval = [7.3397 ns; 7.5151 ns] (CI 95%)
    Skewness = 2.48494238040521, Kurtosis = 9.54399868306873


    Total time: 00:03:47 (227.23 sec)

    // * Summary *

    Host Process Environment Information:
    BenchmarkDotNet.Core=v0.9.9.0
    OS=Windows
    Processor=?, ProcessorCount=4
    Frequency=2435885 ticks, Resolution=410.5284 ns, Timer=TSC
    CLR=CORE, Arch=64-bit ? [RyuJIT]
    GC=Concurrent Workstation
    dotnet cli version: 1.0.0-preview3-003206

    Type=Program  Mode=Throughput

              Method |    Median |    StdDev |
    ---------------- |---------- |---------- |
        TestPeekLong | 8.0208 ns | 0.1583 ns |
     TestOldPeekLong | 7.2402 ns | 0.5349 ns |

    // ***** BenchmarkRunner: End *****
    BenchmarkDotNet.Reports.Summary

Test script: https://gist.github.com/halter73/3cd704aad57ece45ee516c0ac897aea5

So the old code is about 29% faster in the boundary case.

**Comment by cesarbs** [comment]
Just checked plaintext and it's not affected. LGTM then. :shipit:

**Comment by halter73** [comment]
`TryPeekLong` is a little faster than the nullable version.

    blockBytes = 0

              Method |    Median |    StdDev |
    ---------------- |---------- |---------- |
        TestPeekLong | 8.8993 ns | 0.9378 ns |
     TestTryPeekLong | 6.7187 ns | 0.5118 ns |
     TestOldPeekLong | 7.8253 ns | 1.0115 ns |

    blockBytes = 8

              Method |    Median |    StdDev |
    ---------------- |---------- |---------- |
        TestPeekLong | 7.1544 ns | 0.4874 ns |
     TestTryPeekLong | 5.2740 ns | 0.6020 ns |
     TestOldPeekLong | 4.7130 ns | 0.4095 ns |

File: test/Microsoft.AspNetCore.Server.KestrelTests/MemoryPoolIteratorTests.cs [hunk 3]
```diff
@@ -406,10 +404,9 @@ public void PeekLongAtBlockBoundarayWithMostSignificatBitsSet(int blockBytes)
             var scan = block.GetIterator();
             var originalIndex = scan.Index;
 
-            // Act
-            var result = scan.PeekLong();
-
             // Assert
+            ulong result;
+            Assert.True(scan.TryPeekLong(out result));
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/MemoryPoolIteratorTests.cs]: I hadn't thought of this before, but you should add tests for the failure cases.
